### PR TITLE
fix: division by 256

### LINKED ISSUE #20: Typo?
Hi! I've been watching your open source maintenance stream, and noticed you've been talking about diving numbers by 256. 
But this code does not do it:

https://github.com/jonhoo/ordsearch/blob/8eb3b2ef21edd187ea920afea17480fc399b243d/benches/search_comparison.rs#L237-L241

I think there are rarely cases when we want to _divide_ and _multiple_ by the same number and there are certainly none for `usize`s (unless the dividend is smaller then 16, but I think in that case it would've been better to express it more explicitly). 
Not sure if that's important or not, just wanted to point that out.

### PR DESCRIPTION
**Pull request by alisa101rs**
Fixes #20

### CODE REVIEW

File: benches/search_comparison.rs [hunk 1]
```diff
@@ -235,9 +235,7 @@ fn construction_bench_case<const MAX: usize, T, Coll>(
         for e in v.iter_mut() {
             r = r.wrapping_mul(1664525).wrapping_add(1013904223);
             if duplicates {
-                let r = (r % size) / 16 * 16;
-                let r = std::cmp::min(r, MAX);
-                *e = T::try_from(r).unwrap();
+                *e = T::try_from(r % 256).unwrap();
```
Comment (bazhenov) [inline benches/search_comparison.rs]: I don't think this solution is correct. The `/ 16 * 16` part is correct IMO – it is required to provide duplicates (maybe it would be more concise to write as clearing LSB ` & ~0xF`). I think we need to modulo divide by `MAX` which is supposed to be the maximum value of the generated element in the collection.

Comment (alisa101rs) [reply benches/search_comparison.rs]: I am sorry I don't really understand this code 😅 I am going in blind and just doing what @jonhoo or you are saying.

I fixed the code according to your comment, please take a look

Comment (bazhenov) [reply benches/search_comparison.rs]: Don't worry, I don't understand it fully either 😅. I'm pretty sure we will be able to find a working solution together.

Speaking of uncertainty. It's indeed somewhat hard to reason about what numbers does this code produce. What do you think of isolating the random number generating code in the separate iterator, so it can be tested and easily reused?

**Comment by alisa101rs** [comment]
Yeah, actually, now when I look at code, I can see that this 

https://github.com/jonhoo/ordsearch/pull/21/files#diff-d4363780b78d19e382fc015f5711e5ea0342edb9721b42de7b03e7e052611cdbR219-R233

lines aren't really doing anything (other then specifying length of the future data vector. 

I can replace it with something else

**Comment by alisa101rs** [comment]
How about something like that?

File: benches/search_comparison.rs (hunk 1 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: I refactored it to separate function, but I am not sure if you want it to be put in the `lib.rs` so it could be present in the unit tests? Code under `benches` directory cannot be part of unit test.

Comment (alisa101rs) [reply benches/search_comparison.rs]: I noticed that code before had some unused trait bounds. And also there were logically dead code (see my comment in the PR) which I also removed.

Comment (alisa101rs) [reply benches/search_comparison.rs]: Also tests under `"Construction (with duplicates)"` were passing wrong flag to the bench function. And sequence generator itself did not really produce duplicated sequences, tried to address it too.

File: benches/search_comparison.rs [hunk 2]
```diff
@@ -285,3 +240,64 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    let mut index = 0;
+    std::iter::from_fn(move || {
+        let item = if duplicates {
+            let item = std::cmp::min(index, max);
+            item % 256
+        } else {
+            std::cmp::min(index * 2, max)
```
Comment (alisa101rs) [inline benches/search_comparison.rs]: For `u8` the sequence will be:
```
0 2 4 6 .. 254 256 256 256 256 256 ... 256
```
Which doesn't really make sense to call it `duplicates = false`?

Comment (bazhenov) [reply benches/search_comparison.rs]: Indeed, we cannot generate more than 128 values for `u8` without duplicates (only even values). I think it is implemented in such a general way to get rid of duplicated code.

File: benches/search_comparison.rs [hunk 3]
```diff
@@ -285,3 +240,64 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+/// assert_eq!(
+///     pseudorandom_iter::<u8>(16, 0, false, u8::MAX).take(16).collect(),
+///     vec![30, 4, 18, 8, 6, 12, 26, 16, 14, 20, 2, 24, 22, 28, 10, 0]
+/// )
+/// ```
+///
+/// With duplicates:
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u32>(16, 0, true, u32::MAX).take(16).collect(),
```
Comment (bazhenov) [inline benches/search_comparison.rs]: It seems like we do not need first argument (`size`). We have `take()` instead.

File: benches/search_comparison.rs [hunk 4]
```diff
@@ -285,3 +240,64 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+///     vec![15, 14, 9, 8, 5, 10, 0, 4, 5, 0, 4, 6, 5, 14, 12, 2]
```
Comment (bazhenov) [inline benches/search_comparison.rs]: It seems strange to me that all values are such small considering the fact that `max = 2^32`.

File: benches/search_comparison.rs [hunk 5]
```diff
@@ -285,3 +240,64 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+) -> impl Iterator<Item = T>
+where
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    std::iter::from_fn(move || {
+        // LCG constants from https://en.wikipedia.org/wiki/Numerical_Recipes.
+        seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
+
+        if duplicates {
```
Comment (bazhenov) [inline benches/search_comparison.rs]: There is very simple way how to repeat items in the iterator without imperative code:

```rust
iter.flat_map(|v| std::iter::repeat(v).take(duplicates))
```

This way we can isolate random number generation and duplicating them.

File: benches/search_comparison.rs (hunk 3 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: `size` is required when we want to "guarantee" to have duplicates. See `duplicate` branch.

Maybe it would make sense to change the signature from `size: usize, duplicate: bool` to `bounded: Option<usize>`? Then we could pass `None` for non-duplicate case and `Some(size)` for duplicate case

File: benches/search_comparison.rs (hunk 4 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: Yeah because it depends on a `duplicate` flag. If it's `true` we need to bound numbers by the `size` so we could have some duplicates present

File: benches/search_comparison.rs (hunk 5 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: Okay, makes sense, what number would `duplicates` (perhaps `duplicate_factor` would be better term in that case?) be equal to? Static `2`? or something like `size // 4`?

Comment (bazhenov) [reply benches/search_comparison.rs]: I suggest to leave it equal to 16 (as it is right now).

https://github.com/jonhoo/ordsearch/blob/2b576b9b04d1142813a3d0e91ba9ba4eb01f49f8/benches/search_comparison.rs#L238

We can change it in the future if we need to.

Comment (alisa101rs) [reply benches/search_comparison.rs]: Okay, but that would make tests of duplicates with size <= 16 contain only a single unique number, is that what you want?

Comment (bazhenov) [reply benches/search_comparison.rs]: Yes, this is how it works right now basically. Again this "anomaly" only appears in tests like `Search (with duplicates) u8/ordsearch/8` where there is quite narrow integer type and small collection size.

It is not the best solution. But I believe it should be addressed separately.

File: benches/search_comparison.rs (hunk 3 above)
Comment (bazhenov) [reply benches/search_comparison.rs]: I would argue it is better to regulate all 3 arguments (duplicates, size and values range) as independent parameters of random number generator. IMO it's easier to reason about the resulting collection this way. But I'm open to discussion 

cc @jonhoo, @Qqwy

**Comment by jonhoo** [comment]
The original intent of `duplicates` was, if I remember correctly, to generate 1, 2, 3, 4, 1, 2, 3, 4 when asked to generate 8 values. That's obviously not what the code _does_, but that was the target. I don't feel strongly about exactly how we generate duplicates though — if you think of a strategy that is better, we should go for it!

**Comment by jonhoo** [comment]
I also don't think the API matters all that much in this instance since it's in the benchmarking code, not public API, so I'd go with whatever seems somewhat reasonable :+1:  As in, whatever you choose, it's an easily-changed decision, so we should spend minimal effort on debating it :p

File: benches/search_comparison.rs [hunk 6]
```diff
@@ -165,36 +165,32 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
-                })
+            iter
+                // Repeat each items 16 times
+                .flat_map(|i| std::iter::repeat(i).take(16))
+                .take(size)
                 .collect()
         } else {
-            (0..*i)
-                .map(|int| {
-                    // Generate only even numbers to provide a ~50% hit ratio in the benchmark
```
Comment (jonhoo) [inline benches/search_comparison.rs:186]: I'd like to keep this comment to explain the `i * 2`.

File: benches/search_comparison.rs [hunk 7]
```diff
@@ -165,36 +165,32 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
+
         let v: Vec<T> = if duplicates {
-            (0..*i)
-                .map(|int| {
-                    let int = std::cmp::min(int, MAX);
-                    T::try_from(int % 256).unwrap()
-                })
+            iter
+                // Repeat each items 16 times
+                .flat_map(|i| std::iter::repeat(i).take(16))
```
Comment (jonhoo) [inline benches/search_comparison.rs]: Let's move the 16 here into a constant just to make it a bit more visible.

File: benches/search_comparison.rs [hunk 8]
```diff
@@ -165,36 +165,32 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
+    T: TryFrom<usize> + Ord + Clone,
     <T as TryFrom<usize>>::Error: core::fmt::Debug,
 {
-    group.bench_with_input(BenchmarkId::new(name, i), i, |b, i| {
-        let size = *i;
+    group.bench_with_input(BenchmarkId::new(name, size), size, |b, &size| {
+        // increasing sequence of even numbers, bounded by MAX
+        let iter = (0usize..)
+            .map(|i| std::cmp::min(i * 2, MAX))
+            .map(|i| T::try_from(i).unwrap());
```
Comment (jonhoo) [inline benches/search_comparison.rs]: Something still feels off about this. In particular, once `i * 2` hits `MAX`, then we'll just generate `MAX` over and over again, which also feels wrong?

File: benches/search_comparison.rs [hunk 9]
```diff
@@ -285,3 +255,50 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+    bound: Option<usize>,
+) -> impl Iterator<Item = T>
+where
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    std::iter::from_fn(move || {
+        // LCG constants from https://en.wikipedia.org/wiki/Numerical_Recipes.
+        seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
+        let r = (seed.wrapping_mul(2)) % max;
```
Comment (jonhoo) [inline benches/search_comparison.rs]: I think this line was actually wrong in the original — it shouldn't have `* 2`. The whole point of only generating even numbers _into_ the collection was so that ~50% of lookups would not have an exact match. But if all the lookups are _also_ even numbers, then that's no longer true.

File: benches/search_comparison.rs [hunk 10]
```diff
@@ -285,3 +255,50 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+        let r = if let Some(bound) = bound {
+            r % bound
+        } else {
+            r - r % 2
```
Comment (jonhoo) [inline benches/search_comparison.rs]: Doesn't this mean we could have avoided the `*2` when going into this branch?

File: benches/search_comparison.rs (hunk 8 above)
Comment (jonhoo) [reply benches/search_comparison.rs]: Which makes me think this should also be `% MAX` (like you've changed the iterator to). Although that feels wrong too, because then you might end up with duplicates even `if !duplicates`. So I wonder if this needs to be an assert.

File: benches/search_comparison.rs [hunk 11]
```diff
@@ -285,3 +255,50 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+///
+/// ```rust
+/// assert_eq!(
+///     pseudorandom_iter::<u8>(0, u8::MAX, Some(32)).take(16).collect(),
+///     vec![27, 9, 5, 20, 18, 15, 5, 24, 15, 8, 1, 6, 7, 15, 29, 17]
+/// );
+fn pseudorandom_iter<T>(
+    mut seed: usize,
+    max: usize,
+    bound: Option<usize>,
```
Comment (jonhoo) [inline benches/search_comparison.rs]: I'm pretty confused by the name of this, because it's not _just_ a bound, it also changes whether even/odd numbers are generated. I think we should either give it a name that dictates the use (`for_construction`?), or we should split `bound` and `even_only` like you originally suggested. I know I said getting the nuances of internal code right isn't that important (and I still believe that), but as written I just find this function confusing, which makes it more likely for bugs to sneak in :sweat_smile:

File: benches/search_comparison.rs (hunk 8 above)
Comment (bazhenov) [reply benches/search_comparison.rs]: IMO it should be `% MAX`. Although it can generate some amount of duplicates in case of `!duplicates` this amount should be negligible. At least when `MAX` much greater than number of values generated. It obviously will not work in case of `u8`, but in this case no solution exists :)

File: benches/search_comparison.rs (hunk 10 above)
Comment (bazhenov) [reply benches/search_comparison.rs]: Maybe we can eliminate the whole this if-else altogether? "Bound"-computations are covered by `max` parameter of the function and duplication of values is now handled by the clients of this function. So AFAIU `pseudorandom_iter()` does not need to worry about duplicates anymore.

Comment (alisa101rs) [reply benches/search_comparison.rs:299]: `bound` and `max` are different. 

In search case, for example, for `u32 size=512` we will have `MAX=2^32`, 
- search array `v` is generated from iterator `(0..).map(|i| i * 2)` with will have `1024` as a highest value
- but `pseudorandom_iter` will keep generating you values in the whole 0..2^32 range 
     > [1013904223, 1196828703, 450456268, 2817872298, 2081980138, 2971988253, 2135380753, 987533808]


So, how would you prefer to handle it? 
1. keep `bound` as a input parameter in the `pseudorandom_iter` 
2. do `let x = r.next().unwrap() % T::try_from((size * 2) % max).unwrap();` with `T: Rem<Output=T> `

File: benches/search_comparison.rs (hunk 11 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: Removed `bound` parameter

File: benches/search_comparison.rs (hunk 10 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: removed `if-else` entirely

File: benches/search_comparison.rs (hunk 9 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: Done

File: benches/search_comparison.rs (hunk 8 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: Replaced with `% MAX`

File: benches/search_comparison.rs (hunk 7 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: Moved to constant

File: benches/search_comparison.rs (hunk 6 above)
Comment (alisa101rs) [reply benches/search_comparison.rs:186]: Kept the comment

**Comment by alisa101rs** [comment]
Fixed the comments, @bazhenov could you please re-review?

File: benches/search_comparison.rs [hunk 12]
```diff
@@ -285,3 +257,17 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
...
+where
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    std::iter::from_fn(move || {
+        // LCG constants from https://en.wikipedia.org/wiki/Numerical_Recipes.
+        seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
+        let r = seed % max;
+
+        Some(black_box(T::try_from(r).unwrap()))
```
Comment (bazhenov) [inline benches/search_comparison.rs]: I don't think we need `black_box()` here, because it's quite general code. It's seems like it was just moved when refactoring the code.

**Review by bazhenov — CHANGES_REQUESTED** [review]
Just a small nitpick. Otherwise, this code is ready to be merged 👍.

File: benches/search_comparison.rs [hunk 13]
```diff
@@ -285,3 +257,17 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
 fn search_sorted_vec<'a, T: Ord>(c: &'a Vec<T>, x: T) -> Option<&'a T> {
     c.binary_search(&x).ok().map(|i| &c[i])
 }
+
+fn pseudorandom_iter<T>(mut seed: usize, max: usize) -> impl Iterator<Item = T>
```
Comment (bazhenov) [inline benches/search_comparison.rs]: Let's remove first parameter. Not only the function is always used with `seed = 0`, but on the call site it's very easy to confuse the seed with the minimum value of generated numbers 😀

```rust
pseudorandom_iter(0, MAX)
```

**Review by bazhenov — CHANGES_REQUESTED** [review]

File: benches/search_comparison.rs (hunk 12 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: done

File: benches/search_comparison.rs (hunk 13 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: agree, removed seed

**Review by bazhenov — APPROVED** [review]
Seems good to me. It's always nice to see PR which removes more code than add 🎉.

File: benches/search_comparison.rs [hunk 14]
```diff
@@ -165,36 +166,33 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
+
         let v: Vec<T> = if duplicates {
-            (0..*i)
-                .map(|int| {
-                    let int = std::cmp::min(int, MAX);
-                    T::try_from(int % 256).unwrap()
-                })
+            iter
+                // Repeat each items 16 times
+                .flat_map(|i| std::iter::repeat(i).take(DUPLICATION_FACTOR))
```
Comment (jonhoo) [inline benches/search_comparison.rs:186]: Should we also shuffle the `v` before we pass it to `setup_fn`?

File: benches/search_comparison.rs [hunk 15]
```diff
@@ -165,36 +166,33 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
-                    // Generate only even numbers to provide a ~50% hit ratio in the benchmark
-                    let int = std::cmp::min(int * 2, MAX);
-                    T::try_from(int).unwrap()
-                })
-                .collect()
+            iter.take(size).collect()
         };
-        let mut r = 0usize;
+
+        let mut r = pseudorandom_iter::<T>((size * 2) % MAX);
```
Comment (jonhoo) [inline benches/search_comparison.rs]: I don't think this will do the right thing if `size * 2 >= MAX`. Shouldn't this just pass in `MAX` directly? Or maybe
```rust
std::cmp::min(size * 2, MAX)
```

File: benches/search_comparison.rs [hunk 16]
```diff
@@ -285,3 +257,18 @@ fn make_sorted_vec<T: Ord>(mut v: Vec<T>) -> Vec<T> {
 fn search_sorted_vec<'a, T: Ord>(c: &'a Vec<T>, x: T) -> Option<&'a T> {
     c.binary_search(&x).ok().map(|i| &c[i])
 }
+
+fn pseudorandom_iter<T>(max: usize) -> impl Iterator<Item = T>
+where
+    T: TryFrom<usize>,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    let mut seed = 0usize;
```
Comment (jonhoo) [inline benches/search_comparison.rs]: This just made me realize that the construction benchmarks (and also the search benchmarks) are always benchmarking the _same_ set of numbers. I can't quite decide whether that's a feature (eliminate noise) or a bug (only benchmarking one particular data set).

**Review by jonhoo — CHANGES_REQUESTED** [review]
Thanks for all your efforts here! A few more notes.

File: benches/search_comparison.rs (hunk 14 above)
Comment (alisa101rs) [reply benches/search_comparison.rs:186]: Do you want to pull ’rand’ crate for that?

Comment (bazhenov) [reply benches/search_comparison.rs:186]: > Should we also shuffle the `v` before we pass it to `setup_fn`?

I don't think we should. Here `v` is used as a payload for `OrderedCollection` in the context of benchmarking the search, not the construction. Therefore it is beneficial to keep `v` sorted and to use `OrderedCollection::from_sorted_iter()`.

Actually, now as I'm looking through the code base, it seems like we can remove sorting from `make_this()` as it is doesn't needed.

File: benches/search_comparison.rs (hunk 16 above)
Comment (bazhenov) [reply benches/search_comparison.rs]: It is indeed a good question. I think it is beneficial to save PRNG state between benchmarking iterations. This way the sequence will be random on the level of different iteration but deterministic on the level of whole benchmark.

Comment (alisa101rs) [reply benches/search_comparison.rs]: could you check the code?

File: benches/search_comparison.rs (hunk 15 above)
Comment (alisa101rs) [reply benches/search_comparison.rs]: fixed

File: benches/search_comparison.rs (hunk 14 above)
Comment (alisa101rs) [reply benches/search_comparison.rs:186]: Removed `make_this`

Comment (alisa101rs) [reply benches/search_comparison.rs:186]: Actually I think it's not fair for a ordsearch construct case, other two cases (`btree` and `vec`) are performing sorting:

```rust
// std
impl<T: Ord> FromIterator<T> for BTreeSet<T> {
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> BTreeSet<T> {
        let mut inputs: Vec<_> = iter.into_iter().collect();

        if inputs.is_empty() {
            return BTreeSet::new();
        }

        // use stable sort to preserve the insertion order.
        inputs.sort();
        BTreeSet::from_sorted_iter(inputs.into_iter(), Global)
    }
}
```

Constructing BtreeSet from iterator will also collects iterator into vector again, maybe those benches should be reconsidered..

Comment (bazhenov) [reply benches/search_comparison.rs:186]: Yep 😞, it's actually used not only in search benchmarks, but also in construction. I guess we need to leave it as it is. Sorry about that.

File: benches/search_comparison.rs [hunk 17]
```diff
@@ -263,7 +264,13 @@ where
     T: TryFrom<usize>,
     <T as TryFrom<usize>>::Error: core::fmt::Debug,
 {
-    let mut seed = 0usize;
+    static SEED: AtomicUsize = AtomicUsize::new(0);
+    let mut seed = SEED
+        .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |value| {
+            Some(value.wrapping_mul(1664525).wrapping_add(1013904223))
+        })
```
Comment (jonhoo) [inline benches/search_comparison.rs]: I actually think I would just `fetch_add(1)` here instead. Otherwise, since we're using the same logic to update the seed as we're doing to generate numbers, we'll just end up with all the random numbers being offset by one step across benchmarks. Whereas adding 1 to the seed will lead to a completely different number series.

Comment (alisa101rs) [reply benches/search_comparison.rs]: done

**Review by jonhoo — APPROVED** [review]
Thanks for the time you've spent on this! I'm going to go ahead and merge it, and then we can make any further improvements in follow-up PRs if necessary :+1:
